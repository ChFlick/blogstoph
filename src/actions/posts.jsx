import database from '../firebase/firebase';

export const addPost = (post) => ({
    type: 'ADD_POST',
    post
});

export const startAddPost = (post) => {
    const { title, content, author, date, published } = post;

    return (dispatch, getState) => {
        const visibility = published ? 'public' : 'private';
        return database.ref('posts/' + visibility).push({ title, content, author, date }).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                published,
                ...post
            }));
        });
    };
};

export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});

export const startEditPost = (id, post) => {
    const { title, content, author, date, published } = post;

    return (dispatch, getState) => {
        const prevPostPublished = getState().posts.find((post) => post.id === id).published;
        const visibility = published ? 'public' : 'private';

        if (prevPostPublished !== published) {
            const prevPostVisibility = prevPostPublished ? 'public' : 'private';
            database.ref(`posts/${prevPostVisibility}/${id}`).remove();
        }

        return database.ref(`posts/${visibility}/${id}`).set({ title, content, author, date }).then((ref) => {
            dispatch(editPost(
                id,
                post
            ));
        });
    };
};

export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = () => {
    const addPostFromSnapshot = (published, posts) => (postSnapshot) => {
        posts.push({
            id: postSnapshot.key,
            published: !!published,
            ...postSnapshot.val()
        });
    };

    return (dispatch, getState) => {
        return Promise.all(getPostsValueDBReferenceByAuth(getState)
        ).then(([publicPostsSnapshot, privatePostsSnapshot]) => {
            const posts = [];
            publicPostsSnapshot.forEach(addPostFromSnapshot(true, posts));

            if (privatePostsSnapshot) {
                privatePostsSnapshot.forEach(addPostFromSnapshot(false, posts));
            }

            dispatch(setPosts(posts));
        });
    };
};

function getPostsValueDBReferenceByAuth(getState) {
    return (getState().auth && getState().auth.uid) ?
        [database.ref('posts/public').once('value'), database.ref('posts/private').once('value')] :
        [database.ref('posts/public').once('value')];
}

