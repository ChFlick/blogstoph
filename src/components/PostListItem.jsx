import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const PostListItem = (props) => (
    <Fragment>
        <Link
            className="post-preview"
            to={`post/${props.post.id}`}
        >
            <h3 className="post-preview__title">{props.post.title}</h3>
            <p className="post-preview__subtitle">{props.post.content}</p>
        </Link>
        <span className="post-preview__data">
            Posted by {props.post.author} at {moment(props.post.date).format('DD. MMM YYYY')}
        </span>
    </Fragment>
);

export default PostListItem;
