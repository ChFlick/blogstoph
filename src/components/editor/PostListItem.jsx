import React from 'react';
import { Link } from 'react-router-dom';

const PostListItem = (props) => (
    <div>
        <Link to={`post/${props.post.id}`}>{props.post.title}<br /></Link>
    </div>
);

export default PostListItem;
