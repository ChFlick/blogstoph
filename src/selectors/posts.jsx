const getVisiblePosts = (posts = [], { published, sortBy = 'date' } = {}) => {
    return posts.filter((post) => {
        const publishedMatch = published === undefined || (!!published === post.published);

        return publishedMatch;
    }).sort((a, b) => {
        return b[sortBy] - a[sortBy];
    });
};

const getPostByIdOrTitle = (posts, searchField) => {
    return posts.find((post) =>
        post.id === searchField ||
        post.title.toLowerCase().replace(/ /g, '-') === searchField.toLowerCase());
};

export { getVisiblePosts, getPostByIdOrTitle };
