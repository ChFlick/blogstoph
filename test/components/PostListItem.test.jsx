import React from 'react';
import { shallow } from 'enzyme';

import { publishedPosts } from '../fixtures/posts';
import PostListItem from '../../src/components/PostListItem';

describe('post list', () => {
    test('should render correctly', () => {
        const wrapper = shallow(<PostListItem post={publishedPosts[0]} />);

        expect(wrapper).toMatchSnapshot();
    });
});
