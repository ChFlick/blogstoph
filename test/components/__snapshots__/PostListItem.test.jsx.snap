// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post list should render correctly 1`] = `
<div
  className="post-preview-item"
>
  <Link
    className="post-preview-item__link"
    replace={false}
    to="post/1"
  >
    <h3
      className="post-preview-item__title"
    >
      Some Title
    </h3>
    <p
      className="post-preview-item__subtitle"
    >
      This is a Content of a Post...
    </p>
  </Link>
  <span
    className="post-preview-item__data"
  >
    Posted by 
    Author of 1
     at 
    01. Jan 1970
  </span>
</div>
`;
